subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshotUtil.java

```diff
@@ -26,44 +26,55 @@
 
 public class SnapshotUtil {
 
-    public static Optional<MetadataSnapshot> getMetadataFromChildren(List<? extends FileSystemNode> children, String filePath, int offset, Supplier<Optional<MetadataSnapshot>> noChildFoundResult) {
-        switch (children.size()) {
+    public static Optional<MetadataSnapshot> getMetadataFromChildren(List<? extends FileSystemNode> children, String filePath, int offset, CaseSensitivity caseSensitivity, Supplier<Optional<MetadataSnapshot>> noChildFoundResult) {
+        int numberOfChildren = children.size();
+        switch (numberOfChildren) {
             case 0:
                 return noChildFoundResult.get();
             case 1:
                 FileSystemNode onlyChild = children.get(0);
-                return PathUtil.isChildOfOrThis(filePath, offset, onlyChild.getPathToParent())
-                    ? getSnapshotFromChild(filePath, offset, onlyChild)
+                return PathUtil.isChildOfOrThis(onlyChild.getPathToParent(), filePath, offset, caseSensitivity)
+                    ? getSnapshotFromChild(filePath, offset, onlyChild, caseSensitivity)
                     : noChildFoundResult.get();
             case 2:
                 FileSystemNode firstChild = children.get(0);
                 FileSystemNode secondChild = children.get(1);
-                if (PathUtil.isChildOfOrThis(filePath, offset, firstChild.getPathToParent())) {
-                    return getSnapshotFromChild(filePath, offset, firstChild);
+                if (PathUtil.isChildOfOrThis(firstChild.getPathToParent(), filePath, offset, caseSensitivity)) {
+                    return getSnapshotFromChild(filePath, offset, firstChild, caseSensitivity);
                 }
-                if (PathUtil.isChildOfOrThis(filePath, offset, secondChild.getPathToParent())) {
-                    return getSnapshotFromChild(filePath, offset, secondChild);
+                if (PathUtil.isChildOfOrThis(secondChild.getPathToParent(), filePath, offset, caseSensitivity)) {
+                    return getSnapshotFromChild(filePath, offset, secondChild, caseSensitivity);
                 }
                 return noChildFoundResult.get();
             default:
-                int foundChild = SearchUtil.binarySearch(children, child -> PathUtil.compareToChildOfOrThis(child.getPathToParent(), filePath, offset));
-                return foundChild >= 0
-                    ? getSnapshotFromChild(filePath, offset, children.get(foundChild))
-                    : noChildFoundResult.get();
+                if (numberOfChildren < 10) {
```

Comment:
Magic number should be extracted to constant and documented.

Reply:
Also, should this be perhaps something that `SearchUtil.binarySearch()` does internally? Maybe rename the method, too?

All we should care about here is to find the thing, how that happens should belong to the search method. Also, if this optimization is useful, then it will be useful for other callers of `SearchUtil`, too.

We could add more cases, like `size = 0`, `size = 1`, `2 <= size < 10`, `default`?

Reply:
The optimization here is useful since
- the method to check equality is substantially faster than the comparison method
- we are only interested in whether we find the child or not and don't require an insertion point.

All the other usages of binary search require an insertion point if the entry isn't found, so binary search will always be faster.

We could add the 0/1/2 child special cases to binary search, though I am not sure if that would make things faster. For the current method, we already have the special cases `size = 0`, `size = 1`, `size = 2`, `3 <= size < 10`, `default`.

Reply:
👍Let's just name the number then.

Reply:
I named the number. I added special cases for 0 and 1 elements in `binarySearch` as well, let's see if that has any effect on performance.